### PR TITLE
Centralize implicit extension activation scanning in new extension scanner service

### PR DESCRIPTION
**Pull request by joyceerhl**
For https://github.com/microsoft/vscode/pull/165783#issuecomment-1315357060

> We shall have a single method (service) that extension service calls to get all scanned extensions (local, remote, web) and this method shall scan extensions using respective services and update manifests.

### CODE REVIEW

File: src/vs/workbench/services/extensions/electron-sandbox/electronExtensionScannerService.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+			case ScanLocation.Local:
+				return this._scanLocalExtensions();
+			default:
+				throw new Error('Cannot scan extensions for scan location ' + scanLocation);
+		}
+	}
+
+	private async _scanLocalExtensions() {
+		await this._cachedExtensionScanner.startScanningExtensions();
+		return this._cachedExtensionScanner.scannedExtensions;
```
Comment (sandy081) [inline src/vs/workbench/services/extensions/electron-sandbox/electronExtensionScannerService.ts:37]: I think `CachedExtensionScanner` can be inlined into this file.

File: src/vs/workbench/services/extensions/electron-sandbox/electronExtensionService.ts [hunk 2]
```diff
@@ -539,7 +541,7 @@ export abstract class ElectronExtensionService extends AbstractExtensionService
 			// fetch the remote environment
 			[remoteEnv, remoteExtensions] = await Promise.all([
 				this._remoteAgentService.getEnvironment(),
-				this._remoteAgentService.scanExtensions()
+				this._unifiedExtensionScannerService.scanExtensions(ScanLocation.Remote),
```
Comment (sandy081) [inline src/vs/workbench/services/extensions/electron-sandbox/electronExtensionService.ts:544]: Why is `_unifiedExtensionScannerService` is not used for scanning local extensions?

**Review by sandy081 — CHANGES_REQUESTED** [review]

**Comment by sandy081** [comment]
Making @alexdima as primary reviewer because these changes are in his area.

File: src/vs/workbench/workbench.desktop.main.ts [hunk 3]
```diff
@@ -84,8 +84,8 @@ import 'vs/workbench/services/files/electron-sandbox/elevatedFileService';
 import 'vs/workbench/services/search/electron-sandbox/searchService';
 import 'vs/workbench/services/workingCopy/electron-sandbox/workingCopyHistoryService';
 import 'vs/workbench/services/userDataSync/browser/userDataSyncEnablementService';
+import 'vs/workbench/services/extensions/electron-sandbox/electronExtensionScannerService';
 import 'vs/workbench/services/extensions/electron-sandbox/sandboxExtensionService';
-import 'vs/platform/userDataProfile/electron-sandbox/userDataProfileStorageService';
```
Comment (alexdima) [inline src/vs/workbench/workbench.desktop.main.ts:88]: Is this import removed by accident?

**Review by alexdima — CHANGES_REQUESTED** [review]
@sandy081 I think you are the primary reviewer here, as you are in charge of extension scanning. I also lack the context of what the intent of the refactoring is. If the intent of `IUnifiedExtensionScannerService` is to be a facade for all extension scanning, then I think the refactoring is incomplete.

For example, `CachedExtensionScanner` is now used both by `ElectronExtensionScannerService` and `ElectronExtensionService`. I would have expected that `ElectronExtensionService` uses only `IUnifiedExtensionScannerService`.

If the intent is to create a facade, should all these classes/methods become deprecated and all users onboarded to the new service?
* `CachedExtensionScanner`
* `IRemoteAgentService.scanExtensions`
* `IRemoteAgentService.scanSingleExtension`
* `IWebExtensionsScannerService.scanSystemExtensions`,
* `IWebExtensionsScannerService.scanUserExtensions`
* `IWebExtensionsScannerService.scanExtensionsUnderDevelopment`
* `IWebExtensionsScannerService.scanExistingExtension`
* `IWebExtensionsScannerService.scanMetadata`
* `IWebExtensionsScannerService.scanExtensionManifest`

The only subtlety on my side that I'm aware of is that `CachedExtensionScanner` has two methods/properties: 
* `.startScanningExtensions()` to kick off the scanning process (this happens early)
* `.scannedExtensions` to actually get the scanned extensions (this happens later)

**Comment by sandy081** [comment]
@alexdima I thought you are aware of the context from the previous activation events builder PR. Sorry for assuming it wrongly. Here is the context:

With auto generation of activation events, this logic got distributed into following places:

### Local Extensions:
https://github.com/microsoft/vscode/blob/86ff830c3abd8010524eb6492e76ddb9cc6e728f/src/vs/platform/extensionManagement/common/extensionsScannerService.ts#L618

### Web Builtin Extensions:
https://github.com/microsoft/vscode/blob/86ff830c3abd8010524eb6492e76ddb9cc6e728f/src/vs/workbench/services/extensionManagement/browser/builtinExtensionsScannerService.ts#L78

### Web User Extensions:
https://github.com/microsoft/vscode/blob/86ff830c3abd8010524eb6492e76ddb9cc6e728f/src/vs/workbench/services/extensionManagement/browser/webExtensionsScannerService.ts#L673

### Remote Extensions
https://github.com/microsoft/vscode/blob/86ff830c3abd8010524eb6492e76ddb9cc6e728f/src/vs/workbench/services/remote/common/abstractRemoteAgentService.ts#L91

### Remote Extension
https://github.com/microsoft/vscode/blob/86ff830c3abd8010524eb6492e76ddb9cc6e728f/src/vs/workbench/services/remote/common/abstractRemoteAgentService.ts#L103

IMO this is polluting the source of scanning for the sake of generating the model for runtime extension service (runtime extensions). My recommendation to @joyceerhl is to have a single service/util/facade that extension service can use to get its runtime scanned extensions model instead of distributing this model knowledge to all sources. 

> I think you are the primary reviewer here, as you are in charge of extension scanning.

Sure, I will drive this. But I need your eyes if I miss something in the context of extension service.

> For example, CachedExtensionScanner is now used both by ElectronExtensionScannerService and ElectronExtensionService. I would have expected that ElectronExtensionService uses only IUnifiedExtensionScannerService.

Yes, I already made this comment and requested the change.

> If the intent is to create a facade, should all these classes/methods become deprecated and all users onboarded to the new service?
CachedExtensionScanner - Yes (deprecate by inlining to the unified service)
IRemoteAgentService.scanExtensions (Yes, can be deprecated)
IRemoteAgentService.scanSingleExtension (Yes, can be deprecated)
IWebExtensionsScannerService.scanSystemExtensions, (No)
IWebExtensionsScannerService.scanUserExtensions (No)
IWebExtensionsScannerService.scanExtensionsUnderDevelopment (No)
IWebExtensionsScannerService.scanExistingExtension (No)
IWebExtensionsScannerService.scanMetadata (No)
IWebExtensionsScannerService.scanExtensionManifest (No)

I do not think all these shall be deprecated because, extension management still depends on `IWebExtensionsScannerService`.  The unified service/facade is only meant for extension service. This unified service internally uses above services to scan and in addition, the service will also update implicit activation events.

If it is not clear, I am happy to jump on a call to explain in detail.

**Comment by alexdima** [comment]
@joyceerhl Sandeep and I had a longer discussion today trying to see where we want to go with this. The problems right now are:
* users of `IExtensionManagementService` get extensions where `activationEvents` contains only the ones explicitly written in the JSON because the scanning happens either on the shared process or on the server process. Users of `IExtensionService` get extensions where `activationEvents` contains also implicit activation events. This is not a ship blocker because we can address this later.
* vscode extensions that use the extensions API to reflect on other extensions get the update `activationEvents` which contains also the implicit ones. This is a bit of a ship blocker because if we address the previous point, we might end up changing again the `activationEvents` that extensions can read to no longer contain the implicit ones.

To tackle the first bullet point, we think the best approach would be to handle implicit activation events as an implementation detail inside the `IExtensionService`, and not by updating the `activationEvents` array.

**Comment by alexdima** [comment]
Let's defer this to debt week. We will tackle this in early December.

**Comment by alexdima** [comment]
Doing this via https://github.com/microsoft/vscode/pull/175462

**Pull request closed without merging**